### PR TITLE
Delete the `is_a()` macro

### LINKED ISSUE #3233: Remove the is_a macro
The is_a macro is inefficient and should be removed. We can compare with MapObjectType instead and create similar enum classes for other objects (Panels, EditorGameBase).

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2017-11-25T12:56:05Z
 * owner: gunchleoc
 * the launchpad url was https://bugs.launchpad.net/bugs/1734441

### PR DESCRIPTION
**Pull request by Noordfrees**
Fixes #3233

### CODE REVIEW

File: src/map_io/map_object_saver.cc [hunk 1]
```diff
@@ -98,27 +95,27 @@ Serial MapObjectSaver::register_object(const MapObject& obj) {
 
 	assert(!rec.registered);
 
-	if (dynamic_cast<Flag const*>(&obj)) {
+	if (obj.descr().type() == MapObjectType::FLAG) {
```
Comment (gunchleoc) [inline src/map_io/map_object_saver.cc]: I think we could gain even more performance by sticking the `==` comparisons into a switch statement and then handling the `>=` within its default case.

File: src/wui/buildingwindow.cc [hunk 2]
```diff
@@ -254,28 +254,25 @@ void BuildingWindow::create_capsbuttons(UI::Box* capsbuttons, Widelands::Buildin
 				      });
 			}
 		} else if (upcast(const Widelands::ProductionSite, productionsite, building)) {
-			if (!is_a(Widelands::MilitarySite, productionsite)) {
```
Comment (gunchleoc) [inline src/wui/buildingwindow.cc:257]: Good catch :)

**Review by gunchleoc — APPROVED** [review]
Code LGTM - 1 idea for further improvement.

Not tested yet.

**Comment by gunchleoc** [comment]
Changes LGTM :)

**Review by gunchleoc — APPROVED** [review]
Testing done. This can go in :)

**Comment by hessenfarmer** [comment]
As it is green and approved I'll do the merge

**Comment by matthiakl** [comment]
https://github.com/widelands/widelands/blob/bb310a5aa23f4bea8ae2eb10f7786ef5ef662ad8/src/map_io/map_object_saver.cc#L131

This produces a new compiler warning now:
https://travis-ci.org/github/widelands/widelands/jobs/734710208#L2946

**Comment by Noordfrees** [comment]
https://github.com/widelands/widelands/pull/4350/files#diff-b01d9a572c1dc305a7263ca85f09fb1a03e3705dedb87bae5b077948f4ca1f14 fixes it

**Pull request merged**